### PR TITLE
Give wasm a proper API, add a small example of how to use that library, and add codemirror to the demo site.

### PR DESCRIPTION
**Pull request by coolreader18**
This is what I was sorta referring to in #230. In order to be able to make rustfuncs out of closures, I needed to change `RustObjectKind::RustFunction.function` from `fn(..) -> PyResult` to `Box<(Fn(..) -> PyResult)>`, but I think that will lend itself to even more interesting stuff down the line.

I also did some more semantic things like renaming `bootstrap.js` to `index.js`, and `index.js` to `main.js`, and refactoring the build and release scripts a bit. LMK if this goes against the conventions for this repository.

I might also add closure injection from JS to python tomorrow, but I just wanted to get this PR started.

Also, sorry for the mess of diffs with merging in changes from master, the majority of the changes are in `wasm/` with a bit relating to rust functions in `vm/src/pyobject.rs`.

### CODE REVIEW

File: tests/snippets/dict.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+def dict_eq(d1, d2):
+    return (all(k in d2 and d1[k] == d2[k] for k in d1)
+            and all(k in d1 and d1[k] == d2[k] for k in d2))
+
+
+assert dict_eq(dict(a=2, b=3), {'a': 2, 'b': 3})
+assert dict_eq(dict({'a': 2, 'b': 3}, b=4), {'a': 2, 'b': 4})
```
Comment (windelbouwman) [inline tests/snippets/dict.py]: How is it possible that this file was added? We already have it in the master branch?

File: vm/src/builtins.rs [hunk 2]
```diff
@@ -317,7 +317,15 @@ fn builtin_filter(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
     Ok(vm.ctx.new_list(new_items))
 }
 
-// builtin_format
+fn builtin_format(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
```
Comment (windelbouwman) [inline vm/src/builtins.rs]: This was already added in another PR, what is going wrong here?

File: vm/src/pyobject.rs [hunk 3]
```diff
@@ -452,9 +452,24 @@ impl PyContext {
...
+            PyObjectKind::RustFunction {
+                function: Box::new(function),
+            },
+            self.function_type(),
+        )
+    }
+
+    pub fn new_rustfunc_from_box(
+        &self,
+        function: Box<(Fn(&mut VirtualMachine, PyFuncArgs) -> PyResult)>,
```
Comment (windelbouwman) [inline vm/src/pyobject.rs]: Can you use here `Box<RustPyFunc>`?

**Comment by windelbouwman** [comment]
Please resolve conflicts with this branch first. It seems like the some files are added again. For example the dict.py file.

File: tests/snippets/dict.py (hunk 1 above)
Comment (coolreader18) [reply tests/snippets/dict.py]: Yeah, sorry. I rebased and force pushed, so those shouldn't be there anymore.

File: vm/src/pyobject.rs (hunk 3 above)
Comment (coolreader18) [reply vm/src/pyobject.rs]: Should I make a RustPyFunc trait? For example https://stackoverflow.com/a/26071172/9236675. I would have made it with just a type alias if it were possible, but the solution in the SO question seems kinda weird.

**Comment by coolreader18** [comment]
I think those files are conflicting because there are some changes to them that I made near the end of the previous PR that I didn't pull into this branch, but all the changes that were made **are** there, just moved around. I'm pretty sure I just copy-pasted them.

File: vm/src/pyobject.rs (hunk 3 above)
Comment (coolreader18) [reply vm/src/pyobject.rs]: Alright, I did that, cause it seems much cleaner than a bunch of `Fn(..) -> PyResult` everywhere.

**Comment by coolreader18** [comment]
I'm working on closures now.

**Comment by coolreader18** [comment]
Okay, those conflicting files are annoying and are preventing travis builds from running, so I'll just leave them in the PR and once it's merged could a maintainer then remove them? I don't think there's any way to do it otherwise. Thanks!

**Comment by windelbouwman** [comment]
@coolreader18 The changes you made are good, but the git history is now pretty messy. Could you make two new pull requests? One with the change of RustPyFunc and the other one with the wasm demo page improvements?

**Comment by coolreader18** [comment]
Alright, sounds good.

**Comment by coolreader18** [comment]
Okay, so I actually can't have a trait alias as set up in the code due to rust-lang/rust#55687 not being merged to stable yet (I have nightly set as my default so I didn't notice it locally), so I'm just reverting to using the long `Fn(..) -> ..` name until then.

**Comment by coolreader18** [comment]
@windelbouwman should I just rebase it all onto one commit and make a PR with that?

**Comment by coolreader18** [comment]
Alright, now moved to #245.

**Comment by windelbouwman** [comment]
> @windelbouwman should I just rebase it all onto one commit and make a PR with that?

Yes, that would be good, but I only see your message while it is too late ;)

**Pull request closed without merging**